sklearn/decomposition/nmf.py

```diff
@@ -254,22 +352,257 @@ def _nls_subproblem(V, W, H, tol, max_iter, sigma=0.01, beta=0.1):
     return H, grad, n_iter
 
 
-class ProjectedGradientNMF(BaseEstimator, TransformerMixin):
-    """Non-Negative matrix factorization by Projected Gradient (NMF)
+def _update_projected_gradient_w(X, W, H, tolW, nls_max_iter, alpha, l1_ratio,
+                                 sparseness, beta, eta):
+    """Helper function for _fit_projected_gradient"""
+    n_samples, n_features = X.shape
+    n_components_ = H.shape[0]
+
+    if sparseness is None:
+        Wt, gradW, iterW = _nls_subproblem(X.T, H.T, W.T, tolW, nls_max_iter,
+                                           regul=alpha, l1_ratio=l1_ratio)
+    elif sparseness == 'data':
+        Wt, gradW, iterW = _nls_subproblem(
+            safe_vstack([X.T, np.zeros((1, n_samples))]),
+            safe_vstack([H.T, np.sqrt(beta) * np.ones((1,
+                         n_components_))]),
+            W.T, tolW, nls_max_iter, regul=alpha, l1_ratio=l1_ratio)
+    elif sparseness == 'components':
+        Wt, gradW, iterW = _nls_subproblem(
+            safe_vstack([X.T,
+                         np.zeros((n_components_, n_samples))]),
+            safe_vstack([H.T,
+                         np.sqrt(eta) * np.eye(n_components_)]),
+            W.T, tolW, nls_max_iter, regul=alpha, l1_ratio=l1_ratio)
+
+    return Wt.T, gradW.T, iterW
+
+
+def _update_projected_gradient_h(X, W, H, tolH, nls_max_iter, alpha, l1_ratio,
+                                 sparseness, beta, eta):
+    """Helper function for _fit_projected_gradient"""
+    n_samples, n_features = X.shape
+    n_components_ = W.shape[1]
+
+    if sparseness is None:
+        H, gradH, iterH = _nls_subproblem(X, W, H, tolH, nls_max_iter,
+                                          regul=alpha, l1_ratio=l1_ratio)
+    elif sparseness == 'data':
+        H, gradH, iterH = _nls_subproblem(
+            safe_vstack([X, np.zeros((n_components_, n_features))]),
+            safe_vstack([W,
```

Comment:
PG solver is deprecated, and will be removed in 0.19, so I left it like this.
If we change our mind and decide to keep PG solver, I think this sparseness trick should be removed.
I added temporarily L1 and L2 regularization to PG solver to be more consistent with CD solver.